scipy/special/_support_alternative_backends.py

```diff
@@ -53,25 +55,24 @@ def __f(*args, _f=_f, _xp=xp, **kwargs):
     return __f
 
 
-def _get_shape_dtype(*args, xp):
-    args = xp.broadcast_arrays(*args)
-    shape = args[0].shape
-    dtype = xp.result_type(*args)
-    if xp.isdtype(dtype, 'integral'):
-        dtype = xp.float64
-        args = [xp.asarray(arg, dtype=dtype) for arg in args]
-    return args, shape, dtype
-
-
 def _rel_entr(xp, spx):
     def __rel_entr(x, y, *, xp=xp):
-        args, shape, dtype = _get_shape_dtype(x, y, xp=xp)
-        x, y = args
-        res = xp.full(x.shape, xp.inf, dtype=dtype)
-        res[(x == 0) & (y >= 0)] = xp.asarray(0, dtype=dtype)
-        i = (x > 0) & (y > 0)
-        res[i] = x[i] * (xp.log(x[i]) - xp.log(y[i]))
+        # https://github.com/data-apis/array-api-extra/issues/160
+        mxp = array_namespace(x._meta, y._meta) if is_dask(xp) else xp
+        x, y = xp_broadcast_promote(x, y, force_floating=True, xp=xp)
+        xy_pos = (x > 0) & (y > 0)
+        xy_inf = xp.isinf(x) & xp.isinf(y)
+        res = xpx.apply_where(
+            xy_pos & ~xy_inf,
+            (x, y),
+            # Note: for very large x, this can overflow.
+            lambda x, y: x * (mxp.log(x) - mxp.log(y)),
+            fill_value=xp.inf
+        )
+        res = xpx.at(res)[(x == 0) & (y >= 0)].set(0)
+        res = xpx.at(res)[(x != x) | (y != y) | (xy_pos & xy_inf)].set(xp.nan)
```

Comment:
I'd like to avoid these. Will this work?
```suggestion
        res = xpx.at(res)[xp.isnan(x) | xp.isnan(y) | (xy_pos & xy_inf)].set(xp.nan)
```

Reply:
Right, I knew NumPy would go through the compiled code. I think the `np.errstate` context manager would work with the other backends, but not Dask IIRC. Is it possible for *a* context manager to suppress the Dask warning, even if it doesn't exist yet?

Reply:
Dask would need to implement capturing the np.errstate context and replicate it inside the tasks. It's technically doable, but quite hacky.

Reply:
I ask because warnings - even just from NumPy - have been a point of consternation for a long time. 

Suppose the context manager were to work with all backends. Would you still prefer `apply_where` over a context manager?

Reply:
I think there was an open Dask issue for this and it's a recurring issue, so my impression is that it seems worth doing inside Dask despite it being hacky.

Reply:
dask/dask#3245